### DISCUSSION TITLE
libcurl multi: how to reuse the connection of a stopped transfer

### DISCUSSION THREAD

**Discussion by armot5** [curl/curl#12666] [Q&A] [answered]
In implementing a [DASH](https://en.wikipedia.org/wiki/Dynamic_Adaptive_Streaming_over_HTTP) client, we use libcurl to download the MPD file and the segment files. 

On opening a DASH stream, one curl easy handle is created per media type (audio, video, subtitles), and added to the curl multi handle. When closing the stream, all curl easy handles are removed from the multi handle (perhaps in the middle of a transfer) and then cleaned up.

Because of the TCP congestion algorithm, it takes some time before the throughput of a new TCP connection goes up to the normal (max) value. To shorten the zapping time (the time between closing one stream and displaying the next), we try to reuse the TCP connections of the previous DASH stream as long as the next stream is from the same server.

However the WireShark log shows, very small amount of the TCP connections are reused. On closing a DASH stream, if the transfer is in progress, RST is sent to the remote site and the TCP connection is sort of dead afterwards. And there is little improvement in the zapping time.

Because the size of the HTTP data (a DASH segment) is limited, the transfer should be done in a short time. My question is, is there way to keep such TCP connection still usable after the curl easy handle is removed out of the multi handle and before the transfer is done?

**Comment by bagder** [comment, marked as answer]
When doing an HTTP/1 transfer there is no way to stop the transfer early and still reuse the connection for a subsequent request. You can do that fine with HTTP/2 and HTTP/3.

With HTTP/1, the data is on the connection and it has to be read to completeness before the connection can be used again. If the transfer is aborted early, it cannot be used again. Since the server is still sending data on it and the client cannot make it stop without closing it.

**Reply by armot5** [reply]
The server does not support HTTP/2 or HTTP/3. Is it possible to introduce a curl option for this case, to let libcurl discard the incoming data instead of closing/resetting the connection? Thus the TCP connections can be reused and we get the high throughput at the beginning of the transfer of the new stream.

**Comment by dfandrich** [comment]
The application can accomplish this without a libcurl change by simply throwing out
the data from within the callback if it knows that would be more efficient.

**Comment by armot5** [comment]
Yes, but it would increase the complication of the application code -- it means the resources (e.g. the curl easy handles) need to be kept after the stream playback is closed. And at the beginning of the new stream playback, two set of resources run at the same time, one playing the new stream, the other doing the post-processing after closing the previous playback.

In video playback, jumping forward or backward is very common. We stop the current downloading and start downloading from a new position. That means all the curl connections are reset after jumping. That is a pity.

The curl multi handle is to share the connections among the easy handles. It would be nice to reuse the connections of a easy handle when the easy handle is removed/closed before its transfer is done.

**Reply by jay** [reply]
> It would be nice to reuse the connections of a easy handle when the easy handle is removed/closed before its transfer is done.

@bagder explained in his answer why you can't do that for HTTP/1. It's a protocol limitation not a libcurl limitation. Your application has requested a certain offset of data and the server is in the middle of sending all of that data. You could ask for a smaller amount of data ([CURLOPT_RANGE](https://curl.se/libcurl/c/CURLOPT_RANGE.html)) and do what @dfandrich said and ignore the rest of the data. Neither of those will make the connection instantly reusable.

**Comment by armot5** [comment]
Got it. Thanks for the suggestions.
